### ISSUE TITLE
Commands interface

### ISSUE THREAD

**Issue by Shatur** [neovim/nvim-lspconfig#1937]
We currently have API which allows to set language server specific commands. Example from clangd:
https://github.com/neovim/nvim-lspconfig/blob/c5dae15c0c94703a0565e8ba35a9f5cb96ca7b8a/lua/lspconfig/server_configurations/clangd.lua#L52-L59

But @justinmk [suggested](https://github.com/neovim/nvim-lspconfig/pull/1921#issuecomment-1143705517) to remove them.
I personally find them useful, especially `ClangdSwitchSourceHeader`. I don't think it's good for users to copy them from other people's configs. And having a separate plugin for a few lines of code for `ClangdSwitchSourceHeader` is also weird. So I think having them here is the right place.

So it's more like a feature keep request :)

**Comment by mellery451** [comment]
I very much like and need this specific functionality, but I can understand wanting to keep the core clean. If anything, maybe we can come up with a plan to move this to a community plugin like https://github.com/p00f/clangd_extensions.nvim

**Comment by Shatur** [comment]
There are extensions for other LSP servers, having a separate plugin for some additional commands doesn't sound very convenient. Many people complain about the complexity of nvim-lspconfig, forcing them to install additional plugins to get the basic functionality, like `ClangdSwitchSourceHeader`,  which isn't good.

And I don't understand how the commands feature can be a maintenance issue. It's just a few lines of code and many people maintain the language servers they use. If a command breaks - the people who use it will fix it.

**Comment by justinmk** [comment]
> And I don't understand how the commands feature can be a maintenance issue. It's just a few lines of code

Multiplied by 100 or more. I would like to know if we can auto-complete server extensions instead.

**Comment by Shatur** [comment]
> Multiplied by 100 or more--on top of the other activity already.

Users maintain language servers they add. Why it becomes a problem to maintain 1-2 commands (if they exists, most servers don't have it) for the language server they use? How it different from maintaining the language server configuration?

> which is, again, strictly a set of quickstart configs, not a factory for all things that could possibly be useful for each random langserver.

Commands was always here. Why break people configs? Is there any maintenance problem with this feature? I would consider commands as a part of server configuration (as it was in VSCode when you installing a plugin for specific language).

> I would like to know if we can auto-complete server extensions instead. That's worth having in core.

Not sure how it could work.

**Comment by justinmk** [comment]
Just noticed something (added 8 months ago: https://github.com/neovim/nvim-lspconfig/commit/ba90dc2930594ecaae7f5c088038fb8b540869f4) in CONTRIBUTING.md: https://github.com/neovim/nvim-lspconfig/blob/d57b24c3f46c92cf7e1d93919678565ae49e1e23/CONTRIBUTING.md#scope-of-lspconfig

> The point of lspconfig is to provide the minimal configuration necessary for a server to act in compliance with the language server protocol. In general, if a server requires custom client-side commands or off-spec handlers, then the server configuration should be added without those in lspconfig and receive a dedicated plugin such as nvim-jdtls, nvim-metals, etc.

**Comment by glepnir** [comment]
closed as duplicate #1750 .  I  will implement it later.

**Comment by Shatur** [comment]
@glepnir how it is this related? It's a different issue.

**Comment by glepnir** [comment]
that can support custom command in server settings then you can do it right?

**Comment by Shatur** [comment]
> that can support custom command in server settings then you can do it right?

The issue you mentioned is about custom command handlers, if I understand correctly. This issue is about Neovim commands specific to a server. For example, [rust_analyzer](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#rust_analyzer) (see below in `Commands` section) have `CargoReload`.
And users can define their own commands, for example [here is my config](https://github.com/Shatur/neovim-config/blob/ae8a7ef914d7a1d4ffa36be33aeee8b593930e43/plugin/lsp.lua#L129-L142). But this feature was proposed for removal and I decided to open an issue to ask developers to keep it.

**Comment by glepnir** [comment]
not a command handler. you can custom a command in server setting.then you can use this command in neovim cmdline.

**Comment by Shatur** [comment]
> you can custom a command in server setting.then you can use this command in neovim cmdline.

This is how it currently works as far as I know.

**Comment by glepnir** [comment]
maybe I understand wrong you want the comand clangd switch source header in default? i am on mobile phone.now.sry

**Comment by Shatur** [comment]
> maybe I understand wrong you want the comand clangd switch source header in default?

No, I like how commands works right now.
And I submitted a PR that adds another commands for rust_analyzer, see https://github.com/neovim/nvim-lspconfig/pull/1921
But it was rejected due to "we going to remove the commands feature" (see https://github.com/neovim/nvim-lspconfig/pull/1921#issuecomment-1143705517). And then I opened this issue asking to keep it.

**Comment by glepnir** [comment]
Oh I think I understand.  sorry for closed it before.  I've been away for a while due to physical problems. Missed some progress on lsp in neovim. I've been trying to look at these codes recently. Check out the roadmap with `lspconfig` and some justin's comments. I thought I could explain a little bit about the purpose of this project.  Please correct me if I am wrong @justinmk. I will hide or delete this comment. so as not to confuse others about this project.

lspconfig provides basic settings for all servers. And do some actions around the lsp service such as `start` `shutdown` etc . Why does this project exist. Consider not all users have knowledge of lsp and server settings. We create this project. It is convenient for any neovimer to simply enable the language server. But please note that `lspconfig` should only provide the most basic settings. Now some people understand that if you want to use neovim's lsp, you must use `lspconfig`. Do not. Anyone can make any language server plugin around neovim's lsp api. We would prefer to see some neovim lsp plugins in the future that will focus on setting up one or more language servers. In-depth development to adapt to more users of the language. like `nvim-jtdls` or some other plugins. maybe in future you will see like `nvim-gopls` `nvim-tsserver` such as plugin that do most work in server config. Currently the project is bloated. You can see that a large number of language server configurations exist. Maybe you may only use a few of these servers. So that's why you see some commands or something getting deleted. Because we cannot guarantee that it will be needed by all users of that language (personal opinion) .  But based on some api of `lspconfig` you can fully implement it on your personal configuration.   

At present our main work is to promote the progress of lsp on the neovim core. Development of some features. Currently known backward features such as inlay hint smantictoken are not implemented yet. These will have higher priority. 

cc @bennypowers  #2089 also do same thing , explain in there.

**Comment by Shatur** [comment]
> Currently the project is bloated. You can see that a large number of language server configurations exist.

Yes, there are a lot of configs in the repo... But aren't servers loading lazily?

> So that's why you see some commands or something getting deleted. Because we cannot guarantee that it will be needed by all users of that language (personal opinion) . But based on some api of lspconfig you can fully implement it on your personal configuration.

Of course, I can and have already added the commands I need to my config. But if I understand correctly, then the API for such a command is planned to be removed. So I won't be able to define my custom commands in my config either :(

**Comment by glepnir** [comment]
API can be open. I'll check if it's missing. I'll re-add it if I can. Yes, most people use lspconfig as lazyload use autocmd event.
but there still many server configurations file in your disk. this project still accept the new lspconfig. So it will only get bigger and not smaller. :) Thanks for your unsterstand.

**Comment by glepnir** [comment]
I will think about this then make a pr for it .

**Comment by justinmk** [comment]
Note: this would be a core `vim.lsp` feature.

**Comment by justinmk** [comment]
Tracked in Nvim core at https://github.com/neovim/neovim/issues/28329

**Comment by justinmk** [comment]
Clarification: lspconfig is intended to be pure "data"; and the builtin Nvim LSP client (`vim.lsp`) is the "framework".

I've been thinking about this more, and have relaxed on the idea of putting command wrappers in lspconfig. But we need to make sure that those are minimal, and we don't end up building bespoke plugins for each config. Part of this is tracked in https://github.com/neovim/neovim/issues/28329

**Comment by justinmk** [comment]
# Resolution

The legacy `commands` interface in nvim-lspconfig doesn't actually do much except call `nvim_buf_create_user_command`, and doesn't even eliminate much boilerplate.

Discussion in https://github.com/neovim/neovim/issues/28329 revealed:

- lsp server code-actions are discoverable via `gra` (or `vim.lsp.buf.code_action()`)
    - TODO: should `gra` also include code-lenses?
- lsp doesn't provide any way to programmatically list/discover server-provided "custom" (off-spec) commands (`workspace/executeCommand`)
- The `vim.lsp.commands` interface is for defining client-side hooks which the *server* calls. It isn't for user-interactive usage.

https://github.com/neovim/neovim/issues/28329#issuecomment-2782749254

We have now implemented:

- Guidance for nvim-lspconfig:
    - CI lint check enforces: https://github.com/neovim/nvim-lspconfig/pull/3754
        - Config-defined commands must be buffer-local.
        - Config-defined commands must started with `Lsp` prefix.
    - CONTRIBUTING.md mentions:
        - Do not create commands that just alias to code-actions or code-lenses. Those are already discoverable via `gra` (or `vim.lsp.buf.code_action()`).
